### DISCUSSION TITLE
Using this software - the journey

### DISCUSSION THREAD

**Discussion by oatybiscuit** [bixb922/crank-organ#4] [Show and tell]
Rather than continuing to post 'issues' I am continuing my _I intend to use this myself!_ as a _discussion_ with an update on progress so far.

Hardware is on hold while I gather the necessary components, and I aim to build the controller with solenoids as a test first. Once all is proven I can consider how to go about modifying pipework etc and installing in the organ.

### Using an M5stack Core S3.
A nice unit, and I have managed to get it loaded and connected with the Uiflow tool. I am experimenting with various extension modules  - rotation sensors, pressure sensors, and a midi synth unit. It helps to use the Uiflow with blocky to easily find the 'code' for the extra units and the internal display screen, but I find it less satisfactory for programming than directly with python code. I have added Thonny to my PC and found this a great tool - easy to connect to the S3, easy to navigate the flash, and easy to use the REPL to test code directly.

I have managed to write simple code using the M5 libraries to access rotation, pressure sensors and to write to the screen. I have taken the umidiparser code (see different repository) and written a very simple program to load the necessary modules, initialise the synth unit, add the SD card, read a fixed midi file, and then play the midi file sending the midi note on commands to the synth unit. It works!

The hope is that I can somehow use this CoreS3 for front-end display / control, and to add rotation and pressure, and also possibly use the synth for drums etc or an additional pipe register. It would be interesting to have a small amp-speaker to add sub-octave 'helpers' to the base and accompaniment notes.

Using this for the main organ program itself may not be possible without wiping the Uiflow from firmware, and timing may mean that it is just not possible for example to continually read a pressure sensor to provide a gauge meter. My thought is to consider using one esp32 S3 for the main program, and the M5Stack Core S3 unit for monitoring / control with a connection between them.

### Using other S3 boards
I bought an adafruit Metro ESP32-S3, which works nicely and I have been able to write micropython to the flash using Thonny, however the adafruit board is intended for circuitpython, and I have struggled with getting this to work. I added all the organ data and software files manually, but found running the code caused a string of issues. I assume the existing UM loader is getting in the way.

So off to yet another S3 board, this time Unexpected Maker ProS3. Staring from scratch I have added Python to my PC (tip - make sure you tick the 'add to PATH' option otherwise nothing works). I have used Thonny to add micropython to the board, and I have run mpremote in Python from my PC command window and executed the install commands. (note, the software install takes quite a long time and nothing much visually happens for a while).

### The software is up and running

The organ software works nicely. I have just had to edit the main.py to change the initial NeoPixel pin number for this board. The WiFi works and the board is now configured to connect to my home network, and typing 'organ.local' in a browser finds the organ pages. I have added just the one midi file (to flash) and I am still using the default 20 key config, but it all "works", and I have even found that I can see and play the midi file on the PC!

Next job is to get a breadboard, the MCP23017 expanders, solenoid drivers (I have a 32 channel PNP driver board on order) and wire all this up with real solenoids.

- I need to modify the settings and configuration to use the 26 note option, with MCP23017. I have two expander boards, and plan to chain them on the same I2C bus but with different addresses. I am currently assuming this will be OK since the current config uses two different pairs of pins for two separate I2C busses.
- I need to find a few more midi files to work with
- I am looking at battery-packs, and have found a 12/9/5 volt unit. This is normally 12 volts, which I would use to drive the 12v solenoids which appear to work down to 6 or 7 volts, so an exact 12v is not essential, and I can drive the ESP32 S3 from the regulated 5 volts. The question remains as to if/how I can set up the battery monitoring, if any form of special connection is required?

Once the 'system' is proven I can test the solenoid values that I have, and if they work in the organ, it will be a case of building a new wind chest for the valves and installing everything.

### Questions to ponder

I appreciate that, when a midi file is playing, there is a lot going on and the code has been written to minimise other activities so as not to stop the music. I would like to be able to add wind pressure monitoring, rotation speed monitoring, current tune name display and stop/start controls etc using the M5Stack CoreS3 display. Since the midi tune-player part is just one section of the code with the rest mostly devoted to running the webserver, would it be possible to separate the two, having the tune-player etc on one esp32-S3, and the webserver and any extra code I would like to add on another esp32-S3? I understand that esp32 can communicate between one to another using ESP-NOW. Since these boards are really very cheap, a multi-board system would take the pressure off the midi player and allow extra features on the 'display' end of the system.

**Comment by bixb922** [comment]
Wow, that's quite impressive!

>Using an M5stack Core S3.

Ohh, I have looked at those with envy several times already :-)

>Unexpected Maker ProS3

That seems to be a good quality board, I have read good comments about them in the MicroPython forum.

>I have used Thonny to add micropython to the board,

Thonny seems to be simple and useful. I use VSCode to edit/lint the code and mpremote and old fashioned make files to update the microprocessor. Another nice development environment seems to be Viper IDE https://viper-ide.org

>(note, the software install takes quite a long time and nothing much visually happens for a while).

Oh yes, during that time the software is uploaded via the USB connector.

Rest of the topics: I'll comment that later. Believe it or not: I have to go leave now to play the crank organ at at a Christmas concert :-)

**Comment by bixb922** [comment]
Re: Questions to ponder

>The question remains as to if/how I can set up the battery monitoring, if any form of special connection is required?

What type of battery pack is that, and what type of monitoring are you thinking to do? Batteries are quite difficult to monitor externally. For example, measuring the battery voltage does not lead to good results, since the voltage depends on the current load and discharge curves do not help much there. A power monitor such as the INA220 may be the best option, since you can measure the consumed power directly and then add that over time and compare with nominal battery capacity. But that is somewhat equivalent with what the software currently does indirectly: the software adds up the operating time and the time solenoids are on and correlates that (with a linear multivariable regression) with the user input when the battery is empty to estimate a discharge curve.

>wind pressure monitoring, rotation speed monitoring, current tune name display and stop/start controls

That sounds great! I made some tests with the BMP280 sensor and it seems very adequate for the job. It connects to I2C, so if you put the MCP23017 inside the windchest, you can put this on the same bus (https://www.tapatalk.com/groups/buskerorgan/viewtopic.php?p=7391#p7391).

Rotation speed: do you mean crank rotation speed? There is a capability in the software to measure crank rotation speed both with a simple encoder or with a quadrature encoder. The crank can either only start the tune, or can be set to influence playback speed. To sense rotation, it uses the PCNT hardware on the ESP32-S3 chip, so this has very little overhead. I have had many, many headaches with the crank sensor (mostly problems I introduced myself) and have a writeup pending with the conclusions... At this point I have pending to test a quadrature encoder as crank rotation sensor, which I like a lot, but I managed to connect 12V to the microcontroller input and fry it, and haven't had the courage and time to try again so far (and I needed to order a replacement ESP32-S3 board).

As for tune information display, that's quite interesting!

>Since these boards are really very cheap, a multi-board system would take the pressure off the midi player and allow extra features on the 'display' end of the system.

I never thought about separating the code in two, so this will need some coding and design. I can certainly assist there. On the other hand, the "optimization" while playing back music allows to do tasks while the music plays, but you have to be a bit careful. The main issue is responsiveness, i.e. no task should "hog" the CPU for a long time, since that will mean a delay in the notes that are playing. A small delay, such as 10 to 30 milliseconds, is barely noticeable, since the MIDI player catches up with the tempo if that happens. So if a task needs little CPU time once in a while, it can perfectly run while the music is playing. For example, what I'm currently doing while music is playing:
* Read the crank sensor 5 times a second and calculates the playback speed
* Read the touch button 10 times a second to determine an action
* Flash leds
* Respond to web client requests. The requests ship data back to the client in JSON format. Serving such a request will take, say, 5 or 10 millisec at most. For example, every 2 seconds or so, the browser client requests the progress, and there is no noticeable interference. However, if you navigate, for example, to the ```diag.html``` page while playing music, notes will pile up and then play at once, and the music resumes normally, as an example of the hearable effect of CPU hogs.
* And many other tasks, I think there may be about 20 tasks running

Internally, this multitasking is not done with threads but with a mechanism called ```asyncio``` in Python. You can create a task with ```asyncio.create_task()``` and use the ```await``` statement to yield control to the next task or to put the current task task to "sleep" for some specified time.

If a longer computation is needed, exceeding, say, 30 milliseconds, then this software provides a ```RequestSlice``` call:
```
async with RequestSlice("my sensor", 200):
    do something that could need the CPU for up to 200 milliseconds
````
What this will do is to put the enclosed statements to wait until there is a 200 millisecond (or whatever time you specify) pause or long notes in the music. Music has lots and lots of pauses or long notes! That way, long tasks go absolutely unnoticed because they are scheduled when the music allows it. For example, MicroPython's garbage collection, which used to need up to 150 milliseconds, is handled this way and never interfered with the music.

The only drawback of ```RequestSlice()``` is that the start time for a task to be scheduled is a bit variable. Well, with the ESP32 family, having exactly scheduled tasks (for example every second +- 1 millisecond) is not feasible anyhow, and I don't think this is a restriction at all, especially if the tasks relate to human interaction, where 0.1 to 0.5 second delays are unnoticed or quite acceptable. There are always workarounds for this restriction, I think.

And, for me, the other drawback was that I had to learn about ```asyncio```, but there is a good tutorial here: https://github.com/peterhinch/micropython-async

So as a first go, I would take the easier way and launch a task using ```asyncio.create_task()``` at startup time, and make the task with a ```while True``` loop, ```await``` some time and do the needed stuff there using ```RequestSlice()``` for CPU hogs. ```poweroff.py``` is a example that is not very complex and perhaps easier to follow.

Of the enhancements you mention, the display is probably the most CPU consuming one. It would be interesting to measure the display update time. I  would guess update time is quite fast, probably much less than 200 milliseconds.

I think this route is less work than splitting the software in two. The music player is currently not stressed at all, the response can be monitored  installing a ```asyncio```profiler (```aioprof```) in the microcontroller, which is already recognized by the software. 

But I think this could be the best option: when using a second microcontroller for example for a display and buttons, there are already many JSON webservices coded such as ```/get_progress``` and the tune information (tunelib.json) with information about the current tune being played, and that can be called safely at any with a http client such as ```aiohttp``` to extract the information. All buttons of the user interface such as "Next" or "Da capo" or "Shuffle" are also available as JSON web services, and can be called externally from any client. That's a natural interface of the software, and wouldn't even need to touch the current software, since user interface is done with Javascript in the browser via these same web services. The software will work correctly with two clients (browser and a second ESP32-S3).

To communicate two ESP32-S3, ESPNOW seems to be the way to go, although to use the available web services you would need TCP/IP over WiFi (i.e. the builtin ```network``` module) plus ```aiohttp``` or ```requests```as client, not ESPNOW.  Update: On a second thought, this might be the easiest option to integrate another microcontroller.

Please let me know if this is useful.

**Reply by oatybiscuit** [reply]
This is all _very_ helpful thank you. It provides much food for thought. This is very much an evolving project, and as always applying a bit of thinking and testing of ideas up front helps greatly.

I like the _concept_ of using a basic EPS32-S3 just to run the software (as is) close to the solenoid windchest, and to have a second ESP32-S3 up on top to provide (extra) 'monitoring and control'. The M5Stack CoreS3 appears ideal, since it has a touch-screen, SD Card, power options and a small battery, and plenty of easy to use ports.

Splitting software into parts is never easy, and the overhead of adding in inter-code communication and duplicated global variables and modules is likely to outweigh the benefits. I would prefer to keep your code as it stands, but add in 'extra' features on a second ESP32. I have no experience with asynchronous coding, so this will take me time to get to grips with how it works before I can consider making changes. To add 'extra work' to the existing code, such as monitoring pressure, I think that I first need to list out your 20 or so tasks, and see if anything on the list can be turned off, moved to another board, modified for different or less use, and what room there is for adding in other bits. I need to go away and study `asyncio` and your code!

If I was starting from scratch (which I am not) having got your umidiparser to work with a synth unit, I could see the benefit of just having one ESP board to run a basic 'midi player' from a given file. However, as this still requires a file to play and all the solenoid support, as well as monitoring, garbage collection and so on, a better 'cleave point' is more likely to be with the webserver, which naturally separates the working operation part from the less used and end-user interaction for just monitoring and control.

**Battery** - I am looking at a possible unit with 12, 9 and 5 volt supply. The 12v is nominal and unregulated, but perfect for 12v solenoids. The 5v is perfect for the electronics, and the 9v could run the solenoids or the M5 CoreS3. The battery unit has a simple 'press and count the leds' monitor, but nothing to easily integrate with. I had not looked that closely at your code, but good to know that it does the hard work of estimating use based on solenoid use-time. I have just received a set of 12v solenoid valves that I want to use for testing. _Assuming_ they are suitable, the solenoid pulls in at around 7-8v (so could run on 9v) and apparently consumes just 100mA at 12v, so at 9000 mAh capacity the battery will almost certainly last longer than my arm. Perhaps battery monitoring is just not required!

**Tachometer** - my experimentation has been with a rotary encoder unit for the M5 CoreS3. This uses the I2C bus and has a pulse encoder chip built in, hence it keeps count as it is rotated and the running count can be read or reset back to zero. I was going to look at cannibalizing this and trying to add it to the drive wheel between crank and roll drum, then using the M5Stack to read at intervals. With a known interval of time, and the difference between the current and last reading, I should get 'rotations' per minute. This could run entirely separately on the second board, and perhaps offer RPM directly back to the main board via ESPNOW. Since the unit keeps its own count (at 30 increments per revolution) at the end of playing a roll, the count would be a 'total figure' of how much cranking was required for that particular piece, with modifications to your code for resetting the counter to zero before starting to play a new file. Much room for experimentation.

**Pressure** - again experimenting with the M5 stack BPS unit has proven the possibility of adding this easily to the M5 unit. These sensors (BPM280 and others) are indeed very sensitive, and this would be the easier bit of coding to do. I am unsure this would add any real benefit, since on my organ the reservoir is most often full and exhausting, however it might be interesting to record the crank rotation, reservoir pressure, and open pipes (solenoids) to model organ windage. Perhaps when I have nothing better to do!

**Tune Display** - from a basic point of view, just having the name of the tune on the M5 CoreS3 display on the top of the organ would be useful. This could come from the main board (again perhaps using ESPNOW) as the tune starts. From there I could consider having a bigger display on the organ front. Everything is moving to e-paper, but these units seem to require frequent refreshing on a weekly basis or they 'die'. LED display would use more power, but could be just turned on for 20 seconds at the start of the tune, and all of this could be managed by the second board.

Lots to think about. I need to go away and look more closely at your code. Now I have a working ESP32-S3, some solenoids, and a couple of MCP23017 boards. Just waiting for a 32 channel driver board to arrive, and then it will be time to wire it all together and change the config to the 26 note Universal. When I have solenoids 'playing' a tune, I can begin to think about other things!

Thanks again for your detailed information.

**Comment by bixb922** [comment]
>The M5Stack CoreS3 appears ideal, since it has a touch-screen, SD Card, power options and a small battery, and plenty of easy to use ports.

It's a nice board! Fortunately there are several good quality providers for ESP32-S3 boards, that's good for us!

>a better 'cleave point' is more likely to be with the webserver,

I agree. All functions are exposed as web services, so adding a second user interface should be quite feasible. (edit) For example, to implement a crank rotation sensor in a second microcontroller, there is a /set_velocity_relative web service. I did not document these web services, I never thought somebody might use those :-) but please post your questions here and I'll be happy to find some way to implement that. 

>Perhaps battery monitoring is just not required!

I did all the support for calculating the battery usage, but then found out that at less than 300 grams per battery it's easier to carry a second battery pack around :-) and changing it when required. But it's nice to have a better indicator than just four leds.

>however it might be interesting to record the crank rotation, reservoir pressure, and open pipes (solenoids) to model organ windage

Yes, I want to do that sometime also! There are some excel sheets out there to calculate the theoretical wind usage based on pipe measurements, but measuring the real thing is always interesting and sometimes surprising!

>Tachometer
That should work. That rotary encoder seems to be a quadrature encoder (i.e. counts positive in one direction and negative in the other). Class TachoDriver in tachometer.py uses the Encoder class which in turn uses the low level driver at pcnt.py. It should be feasible to replace the Encoder class used there with an Encoder class that reads the rotary encoder's count via I2C. It's perfectly ok to use MicroPythons SoftI2C to do the I2C management. IDK if M5Stack provides some MicroPython example or some module.

> With a known interval of time, and the difference between the current and last reading, I should get 'rotations' per minute

That's what tachometer.py does currently. Probably a good part of that code can be reused.

>Everything is moving to e-paper

Oh yes, that's one of my dreams in the future, to have a e-paper display for the current tune title/genre/author/year/info. I find the flickering refresh and refresh time (sometimes 10 seconds or so) of many units annoying, but good, fast and non-flickering e-paper displays exist (but can be quite expensive). There are even some bendable e-paper displays out there :-)

**Comment by oatybiscuit** [comment]
Proof of concept - first stage - completed successfully.

ESP32 S3 (Dev Kit equivalent) with breakout board. Updated Micro Python to access 16Mb flash. (separately tested adding SD card and _finally_ got that to work too). All code added and working.

Battery pack (12v and 5v). Two MCP23017 breakout boards (one I2C line, two different addresses). PCL signal controller (32 channel) with 3.3v input signal to control 12v relays. Valve relays (7 wired in for testing).

Playing a basic Midi file.


https://github.com/user-attachments/assets/87caaa62-31bf-46b9-84f1-617963abc7d6

**Comment by bixb922** [comment]
I like that!! Neat setup, looks way better than my breadboards!!! I am very happy that you could make the whole setup work!

**Comment by oatybiscuit** [comment]
I am very happy that it was, relatively speaking, remarkably easy to build a working Midi organ with so little effort. Your software is comprehensive and easy to use.
The electronics are really very basic, but I have decided to go with 'industrial' units. This costs more, but the large 'signal controller' comes in a frame that will be easy to mount, and easy enough to add mounting boards for all the other parts. All parts should be replaceable, all parts have power led, so easy to see what is on, and the signal controller provides full isolation and an indicator led (great for testing).

I am planning on adding a synth unit to provide additional support for bass pipes, drum sounds, and perhaps an additional reed/trumpet rank. I have a couple of SAM2695 units (and amp / speakers). I just need a Midi feed for this, which could come from the organ software, and that would then mean I can fully test with sound before I start on the physical stuff. Next step is to break into the pipework somewhere, put in one of the solenoid valves, and check to see if it actually works and lets enough air to the puffs.

Slow and steady progress.

**Comment by oatybiscuit** [comment]
Update on (lack of) progress.

The valve part of the solenoid-valve was _not_ suitable and did not pass sufficient air to operate the puffs. I have sourced an alternative that does work. This required a rethink around mounting, and I have gone for making a new bleed-bar, incorporating the solenoid/valve, a slider bar to switch the operation from paper roll play to midi play, as well as a set of pressure switches operating at around 2 inches water. These pressure switches took a _very_ long time to source, and then more testing to check that it actually works by building a prototype.

I have re-sourced smaller and lighter hardware to drive the solenoids, and a better battery pack that has regulated 5v, and unregulated 6, 9 and 12v, as well as a state of charge lcd display.

The 'digital ranks' project has gone with Zynthian running on a Raspberry Pi with a DAC amplifier. I did source a keyboard to midi encoder, but for greater flexibility have switched to using an Adafruit Metro (Grand Central). With a bit of Micro Python I have tested a matrix keyboard arrangement to read the pressure switches, with the code generating midi on/off passed to Zynthian over USB. Using Aeolus organ, I have extra 'buttons' for organ stops, so I can bring in a trumpet stop, 32 & 16 foot bass helpers, and a full church organ sound.

The new bleed bar, without the solenoids, is currently being installed, and the pipework extended and modified. This should bring the organ back to working as it was, then the ESP32, new driver boards and the solenoids can be added. The pressure switches need to be wired first, but that will then be left for the next stage once the midi-player is working.

**Comment by bixb922** [comment]
>The valve part of the solenoid-valve was not suitable

I am sorry to hear that. But tinkering is nice in itself, I hope you are enjoying it!

>The 'digital ranks' project....

Wow!!! Sounds great!!! Do you have some pictures or videos somewhere?

**Reply by oatybiscuit** [reply]
This project is mostly _all_ tinkering! Now I have started to actually install the modifications, which means I have to get it right otherwise I might ruin a perfectly good and working instrument.

This is the driver tray that will sit on the top of the valve chest, behind and beneath the new bleed bar. Dual MCP23017 board and four 8-channel drivers (all screw-terminal). The solenoids have plug tails, so the drivers are likewise terminated. The ESP32 S3 board will go somewhere on the back of the valve chest. Just needs 5v to the ESP and 9v for the solenoids. All tested and working. Microphone, a rotary/crank encoder, touch pad/switch, and separate neopixel led on the outer case to be wired in afterwards.

<img width="3729" height="2341" alt="midi player esp32 and solenoid drivers" src="https://github.com/user-attachments/assets/6dd27d1b-f7c9-4217-a3ec-34083d792e36" />

The digital ranks project has got as far as a working prototype, and I am going ahead with fitting the pressure switches to the bleed bar.

The hardware consists of a Raspberry Pi 4 on which runs Zynthian (headless). This requires a DAC amplifier, and I almost gave up entirely, having such difficulty getting anything out of Zynthian on a cheaper test DAC. The secret is the expensive HiFiBerry Amp4 Pro, DAC amplifier. This plugs into the Pi GPIO pins, accepts 12-18 volts, and drives the amp as well as providing regulated 5v for the Raspberry Pi, with enough power to run extra hardware via the USB. Zynthian accepts midi over USB, and I have loaded the Aeolus Organ engine, which is a full three manual and peddle-board church organ based on algorithms rather than synth.

<img width="4017" height="2349" alt="digital pipes - Pi with Zynthian and Metro m4" src="https://github.com/user-attachments/assets/54aa11b0-299f-4a41-bb98-a506ad670ad7" />

The Adafruit M4 Grand Central takes power from the Pi USB. Using Circuit Python there are libraries for midi, midi over USB, and matrix keyboards, so I have a small program that loops round reading a matrix keyboard (which will be the pressure switches, and organ stops). This will send a midi on/off over USB to Zynthian, which then plays the organ note. Stop changes require a complex sequence of command codes, but do operate from simple switches, either momentary or pull on / push off. I will also add a microswitch to the pressure box lid, which will act as a cancel, sending all-notes-off command to Zynthian and cancelling all stops, when the box lid is opened.

It works in test.

The audio will be the challenging part. Simple 3 Watt speakers for testing are almost certainly not going to be loud enough, so I have obtained a couple of exciter-drivers, and will, once the organ is working again, test out the pressure switches and then Zynthian to see if I can get sufficient volume. Perhaps attaching the exciters directly to the organ case. The HiFiBerry DAC will power 20 W, as long as I can get a speaker into the case somewhere.

As an extra I will code in a neopixel LED strip to operate with the notes, so this will provide a view of the action, as well as providing a facility for testing and checking note playing.

If it all works (yet to be proven) there are potential benefits. I can execute Zynthian commands using midi, so an extra set of buttons in the keyboard matrix can be used to start and stop midi recording. This permits a paper roll play to be midi-captured to file, where I can edit and transfer to the midi player. I have the paper rolls, but should be able to re-play my own cranked performance from midi, which will save weight and space for public performances. Yes, it would be nice to extend the neopixel strip to a matrix, and thus add a digital 'paper roll over tracker bar' window to still look like a paper roll player.

Got to get it all working first though!

**Comment by oatybiscuit** [comment]
First part installed, and working.

<img width="3821" height="2749" alt="new bleed bar with solenoid-valves" src="https://github.com/user-attachments/assets/d5d7b885-f924-4f66-a2af-35be6c5be65b" />

New (much larger) bleed bar, with new mounting and hole in side wall for roll-to-midi switching slider.
Pipework extended (over 1.5m additional tubing) and re-routed. Base pipework routed through new bleed bar, with just those new bleed holes closed.
Driver tray loose placed behind bleed bar.

Work in progress:

Tidy solenoid cables if possible.
Secure driver tray - Velcro pads?
Mount ESP32 board and re-wire connection to drivers.
Add battery box to exterior, and wire 5v & 9v power with fuses, switches, power leds.
Add microphone, temp encoder / switch, external Neopixel led.
Add pressure switches to bar and test.

The bleed bar was set to match the lower bar-valve pipework, which requires no modification and is easy to fit. The tracker-bar pipework has less room, required extending, and has little room to work. Very fiddly to get the pipe into the bleed bar over the top of the solenoids.

The valves all work as expected. The only issue is with the bleed screws. These of course had to be reset for manual roll play, and when switching over to midi play almost no notes sounded, as the air pathway via the solenoid-valve is more restricted than directly via the holes in the aluminium bar. Not enough air to open the valve. Problem solved by re-adjusting and closing almost all the bleed screws.

Since this is not an ideal situation - having to adjust bleed screws when switching from roll-play to midi-play or back, I intend to replace the aluminium bar and re-drill with smaller holes. The thought is, by experimentation, to find a size of hole where the direct path airflow matches the via-valve airflow. Another possible option is to try adding a second adjusting screw that goes into the direct flow pathway either side of the switching bar, so as to reduce the main airflow to match the valve airflow. This could be a long project!

**Comment by bixb922** [comment]
Very interesting! I can relate to the difficulties you are experiencing. Pneumatic system are not easy at all!

**Comment by oatybiscuit** [comment]
I went back to the drawing board and redesigned the bleed bar to include _two_ bleed screws per note. The first (lower) main bleed screw and vent operates all the time, exhausting to the bottom of the bar. The second (upper) screw vents upwards back through the slider bar, and operates only in the roll-play mode. Since roll play will have greater air flow, the second bleed screw acts in addition to the first.

I needed to have a higher level of precision with the holes, so I went with hard maple wood this time, and a slightly larger bar of 60mm depth. The extra vertical space allows for more room to ensure that the insert pipework top and bottom does not cover the side holes to the valve and pressure switch. There are now two holes for each note in the slider bar (air feed downwards and bleed exhaust upwards) with smaller holes and much care to ensure accurate drilling and sufficient sealing around each hole. The slider bar now has felt seal on both sides.

There is less room for the lower pipework, so more adjustments required for the pipes from bar to puffs.

Aside from an issue with failed electrical connections to two of the solenoids, all other notes operate as expected. The note repetition test helps with quickly setting the main bleed screw, and at 50ms on / 50ms off it is easy to find the screw positions for

no-note -> incomplete-note -> **clean-repetition** -> hold-over -> sustain

Some fine adjustment is still required, but the midi-play primary bleed screws are set, and the roll play secondary bleed screw adjustment appears to work as hoped for. I am able to play a midi file with rapid trill notes, and a paper roll with rapid trill notes, and there is almost no difference between them.

A couple of months to get back to where I was in June, but still "progress"!

<img width="4093" height="2813" alt="bleed bar with two screws each note" src="https://github.com/user-attachments/assets/cfb8841a-dbfa-4d08-9100-7021bc3f2d19" />

**Comment by oatybiscuit** [comment]
And finally:

New bleed bar adjusted and installed, along with the driver-tray now secured by Velcro, and the ESP32 board on the valve chest. Controls and extras added (tempo, start button, remote Neopixel, and microphone). Slight issue with the board I am using for the ESP32 with at least one faulty connection between header and screw terminal. Little things like this add almost a whole day of fiddling, taking it all apart again, finding the fault, correcting, and putting it all back together. The pipework is very fiddly and it takes two to three hours to install the bleed bar and reconnect all the pipes, so not something I really want to be doing regularly!

The pressure switches, for the MIDI-out project, are installed, wired and individually tested. Next step is to build a battery holder on the side of the organ, then look at adding the Zynthian and Pi for the digital pipes.

The solenoids are 12v but I am running them from the 9v (9-10v unregulated) and they do still make quite a clicking noise. I took one valve apart to see if I could add felt cushioning to stop the noise, but the throw is so small that anything added prevented the valve from opening correctly. The clatter of tiny solenoids will have to add to the charm.

Great software - thank you for writing this. The tuner is very easy to use and tells me that I have 11 pipes to tune. Only issue at the moment is, having added a remote Neopixel led and changed pin, the on-board one still lights up blue. I don't know if this is because the led is still 'programmed' from the last use (although I made sure it was not flashing blue when I changed the pin configuration) or because the program starts on the old 48 pin before switching to the one defined in the pin configuration.

How it looks, and sounds...

<img width="3072" height="4096" alt="IMG_20260831_112252222" src="https://github.com/user-attachments/assets/52781138-b469-4919-8bb3-ba003b828bc6" />

Playing from a MIDI file and not paper roll!

[midi-player.mp3](https://github.com/user-attachments/files/31642551/midi-player.mp3)

**Comment by bixb922** [comment]
That is a very good progress! Thank you for attaching the mp3. I think the sound is wonderful. And Bach is always a treat to hear.

>it takes two to three hours to install the bleed bar 

But it seems to work very well! That must have been lots of work and attention to detail!

>add felt cushioning to stop the noise

I tried that too at one point, but with a 3mm (or even 5mm)  throw, adding a cushion is not possible. I have some small solenoids with a tiny rubber ring, that may be an option? The click can be heard, but as you say it adds some charm. If you really wish to get rid of the click, perhaps you might consider adding casing or a wall so that when hearing the music from the front, the noise is dampened? I am considering that solution for the noise  of the escape valve of the bellows.

I am glad you found the tuner useful. If someone can help tuning, one person turns the crank and the other adjusts the pipe while looking at a tuner (for example a. tuner in the cell phone). But that's not an option for me, so that is how the tuner in the software came into being.

>having added a remote Neopixel led and changed pin, the on-board one still lights up blue

I always connect the external Neopixel to the same pin (e.g. pin 48) than the internal Neopixel. The pin drives both with no problem, and they show the same colors. But there is certainly no problem to use another pin. That said, what you see it is as programmed. On startup, the neopixel on pin 48 always lights up blue, even before reading the configuration file. I wanted to have unconditional feedback that the program started. I'll probably fix that at some point use only the Neopixel in the configuration.

Thank you for posting!
